workflow/airflow/dags/local_file_sensor_dag.py

```diff
@@ -0,0 +1,94 @@
+import os
+
+import airflow
+from airflow import DAG
+from airflow.operators.python_operator import PythonOperator
+from airflow.operators.sensors import BaseSensorOperator
+from datetime import timedelta
+
+
+class FileWatcherOperator(BaseSensorOperator):
+    """
+    Triggers a file system check. Parses a file if present.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize a FileWatcherOperator which should wait for a file to be present in a target directory.
+
+        :param args:
+        :param kwargs:
+        """
+        self.directory_path = kwargs.get('directory_path')
+        super(FileWatcherOperator, self).__init__(*args, **kwargs)
+
+    def poke(self, context, *args, **kwargs):
+        """
+        See if a file is present in a given directory.
+
+        :param context: Context dictionary forwarded from the DAG;
+        expects to contain context.dag and context.task
+        :type context: dict
+        :param args: not expected
+        :param kwargs: not expected
+        :return: whether the poke method has succeeded or not, 0 or 1
+        :rtype: int
+        """
+        path = os.path.join(os.getcwd(), self.directory_path)
+        dir_listing = os.listdir(path)
+
+        if len(dir_listing) > 0:
+            context['ti'].xcom_push(key='file_path', value=os.path.join(path, dir_listing[0]))
+            return True
+
+        return False
+
+
+DIR_PATH = 'data'
+
+
+def parse_file(*args, **kwargs):
+    """Simply parse the first file in a directory and return it's contents.
+
+    :param args:
+    :param kwargs:
+    :return:
+    """
+    # get file path value from passed in context
+    file_path = kwargs['ti'].xcom_pull(task_ids='file_watcher', key='file_path')
```

Comment:
`task_id[s]` seems to be dinamically defined for `FileWatcherOperator`, while here it is hardcoded. What is the standard way to do it?

Reply:
In the `FileWatcherOperator` I am pushing a value onto the store so you don't need to define `task_ids`, it knows what task you are pushing from.

In the `parse_file` function I am saying get me the value for key `file_path` that was populated by the `file_watcher` operator:

`file_path = kwargs['ti'].xcom_pull(task_ids='file_watcher', key='file_path')`

So when pulling values you are required to define a value for `task_ids` otherwise you get:

`TypeError: xcom_pull() missing 1 required positional argument: 'task_ids'`

If you pass `None` to `task_ids` it will just look for the first key `file_path` populated by any of your tasks in that instance (not recommended as could lead to unpredictable results).

If you pass a list of operator names to `task_ids` it will return a tuple of multiple values, in this example, you might have lots of operators pushing file paths and you may want to parse them all at the end (this is contrived but you get what I mean). e.g.

`file_paths = kwargs['ti'].xcom_pull(task_ids=['watcher_1', 'watcher_2', 'watcher_3'], key='file_path')`
```
>>> file_paths
('/path/one', '/path/two', '/path/three')
```

Reply:
> If you pass None to task_ids it will just look for the first key file_path populated by any of your tasks in that instance (not recommended as could lead to unpredictable results).

Unpredictable, true, but if used like the elife-bot's session these values are populated once and never overwritten; so in that use case it could work (but only loosely enforced).